### PR TITLE
Add the gem for user authentication with Google

### PR DESCRIPTION
**Pull request by kinyaoo**
#### Summary

Users should log in with their Google account. Adds the OmniAuth Google
OAuth2 gem in order to authenticate with Google.

#### Ticket Link

Github: https://github.com/mickmister/music-sniper-rails/issues/1
Gitlab: https://gitlab.com/mjkochell/music-sniper/issues/101

### CODE REVIEW

**Comment by mickmister** [comment]
Looking really good! I like the improvements added. Let's continue discussion here so it's closer to the code.

File: app/commands/authenticate_user.rb
Comment (mickmister) [inline app/commands/authenticate_user.rb:11]: This code was terrible! :laughing:

File: app/commands/authenticate_user.rb
Comment (mickmister) [inline app/commands/authenticate_user.rb]: Minor suggestion

```suggestion
	return false unless user && password
	
	return user.authenticate(password)
```

**Review by mickmister — COMMENTED** [review]
Great work @kinyaoo! Just one suggestion

Also, Github can lose track of comments when the original commit hash of the comment is changed. Do you mind pushing to your branch when you make updates, rather than rebasing and force-pushing?

File: app/commands/authenticate_user.rb
Comment (mickmister) [reply app/commands/authenticate_user.rb]: Actually this doesn't work for empty strings. First line should be:

```
	return false unless [user, password].all?
```

**Comment by kinyaoo** [comment]
> Great work @kinyaoo! Just one suggestion
> 
> Also, Github can lose track of comments when the original commit hash of the comment is changed. Do you mind pushing to your branch when you make updates, rather than rebasing and force-pushing?

I could it's just because I use to reorganize my commits by rebasing then I have to push force but commits are cleaner this way I think

**Comment by mickmister** [comment]
> I could it's just because I use to reorganize my commits by rebasing then I have to push force but commits are cleaner this way I think

@kinyaoo Sounds good. Sorry about the lack of activity on my part. I will test sometime this week and get this merged.

**Comment by mickmister** [comment]
@kinyaoo I wanted to see the feasibility of a client-side solution, and it turned out to be very few lines of code with https://github.com/anthonyjgrove/react-google-login.

1. The frontend authenticates with Google
2. The frontend sends to the backend a token that Google returned
3. The backend uses the token to get the user's Google-verified email address
4. The user now has an account, and there are no refresh tokens to be stored in the database

For the original purposes of this feature, these steps satisfy the requirement. I think your solution has the potential to provide many more features. Do you know what use cases the solution in this PR provides outside of the requirement?

**Comment by kinyaoo** [comment]
@mickmister 

I haven't tried but according to https://github.com/omniauth/omniauth I think that this PR can be used to authenticate users through multiple strategies.

We just need to follow these steps :

1. Add and configure the corresponding gem from https://github.com/omniauth/omniauth/wiki/List-of-Strategies

2. Add the new provider in the config/initializers/omniauth.rb file

You can merge this if it's good for you and assign me to the next issues for implementing new providers for authentication if you want

**Comment by kinyaoo** [comment]
Hi, could you review the PR ? @mickmister

File: app/commands/authenticate_user.rb
Comment (mickmister) [inline app/commands/authenticate_user.rb]: I have gotten rid of `user.safe_atrributes` in my code that I haven't merged yet. Can you please change this to `as_json` to match the changes I will introduce later?
```suggestion
    { auth_token: auth_token, user: user.as_json }
```

File: test/models/o_auth_access_data_test.rb
Comment (mickmister) [inline test/models/o_auth_access_data_test.rb]: Could you please write a test here or delete the file? What do you think you would want to / be able to test here?

**Review by mickmister — CHANGES_REQUESTED** [review]
Hi @kinyaoo, I have a few questions

Can you tell me about the lifecycle of a o_auth_access_data record? Do we need to refresh the tokens on some unit of time basis? Are they used upon re-login?

Can you provide steps for me to test this? Maybe something that you have done that I can replicate using a UI component?

Also I want to point out that if you rebase, I have to review all of the changes at once again. I understand that you want to reorganize the commits. But if you instead just add more commits onto the PR, it is easier for me to review as I can see the commits that I have not reviewed yet. That would be very helpful!

Thank you!

File: test/models/o_auth_access_data_test.rb
Comment (kinyaoo) [reply test/models/o_auth_access_data_test.rb]: I prefer to delete it for now

**Comment by kinyaoo** [comment]
> Hi @kinyaoo, I have a few questions
> 
> Can you tell me about the lifecycle of a o_auth_access_data record? Do we need to refresh the tokens on some unit of time basis? Are they used upon re-login?

Refresh token will never change there is only the access token every 2 hours I think. You can easily re-login your user.

> Can you provide steps for me to test this? Maybe something that you have done that I can replicate using a UI component?

Using a UI component you juste have to go on the following URL http://localhost:3000/auth/google_oauth2

> Also I want to point out that if you rebase, I have to review all of the changes at once again. I understand that you want to reorganize the commits. But if you instead just add more commits onto the PR, it is easier for me to review as I can see the commits that I have not reviewed yet. That would be very helpful!

Ok I understand

File: app/models/update_o_auth_access_data_form.rb
Comment (mickmister) [inline app/models/update_o_auth_access_data_form.rb]: Can you help me understand why this code has its own model, as opposed to living in the `OAuthAccessData` class, or something else e.g. a service?

File: app/controllers/omniauth_sessions_controller.rb
Comment (mickmister) [inline app/controllers/omniauth_sessions_controller.rb]: Can we return a useful error instead of the two json responses here? This is shown as raw data in the user's browser after they authenticate. Maybe something like `Connected (email) account` and `Error authenticating`.

**Review by mickmister — CHANGES_REQUESTED** [review]
I've gotten this PR up to speed with master. I've confirmed the feature works correctly. Great job @kinyaoo! :tada:

I've left two comments for feedback. Once they are addressed, we will merge this into master.

File: app/models/update_o_auth_access_data_form.rb
Comment (kinyaoo) [reply app/models/update_o_auth_access_data_form.rb]: This article will explain better than me why I use form objects.
But I agree with you it can be confusing to let this code in the models directory.
We could move this code under a `forms` directory if you prefer ?

File: app/controllers/omniauth_sessions_controller.rb
Comment (kinyaoo) [reply app/controllers/omniauth_sessions_controller.rb]: Something like :
```
if command success?
 render json: "Connected #{command[:user][:email] account"
else
 unauthorized(errors: "Error authenticating")
end
```

**Comment by kinyaoo** [comment]
> I've gotten this PR up to speed with master. I've confirmed the feature works correctly. Great job @kinyaoo! 

Thank you it's a pleasure @mickmister

File: app/models/update_o_auth_access_data_form.rb
Comment (mickmister) [reply app/models/update_o_auth_access_data_form.rb]: > We could move this code under a forms directory if you prefer?

Sounds good to me!

File: app/controllers/omniauth_sessions_controller.rb
Comment (mickmister) [reply app/controllers/omniauth_sessions_controller.rb]: Instead of json, I think we should return html or text.

```ruby
if command.success?
  render html: "Connected #{command.result[:user]['email']} account. You may now close this browser tab."
else
  render html: "Unable to connect account."
end
```

You will also need a `skip_before_action :force_json` at the top of the file to be able to use html.

Comment (mickmister) [reply app/controllers/omniauth_sessions_controller.rb]: Feel free to use some sort of html here if you have a more elegant idea! 

EDIT: If you do decide to use html, I would keep it simple so you can keep it inline here, or move it to a view.

**Comment by kinyaoo** [comment]
@mickmister I just pushed the last commit, sorry for the delay

**Review by mickmister — APPROVED** [review]
Awesome work!! Thanks for addressing all the feedback @kinyaoo

**Pull request merged**
